pricegraph/src/orderbook.rs

```diff
@@ -0,0 +1,207 @@
+                        .last()
+                        .expect("unexpected empty pair orders collection in map")
+                        .weight;
+                    (sell_node, buy_node, weight)
+                }
+            })
+        }));
+
+        Ok(Orderbook { orders, projection })
+    }
+
+    /// Returns the number of orders in the orderbook.
+    pub fn num_orders(&self) -> usize {
+        self.orders
+            .values()
+            .flat_map(|o| o.values())
+            .map(Vec::len)
+            .sum()
+    }
+
+    /// Detects whether or not a solution can be found by finding negative
+    /// cycles in the projection graph starting from the fee token.
+    ///
+    /// Conceptually, a negative cycle is a trading path starting and ending at
+    /// a token (going through an arbitrary number of other distinct tokens)
+    /// where the total weight is less than `0`, i.e. the effective sell price
+    /// is less than `1`. This means that there is a price overlap along this
+    /// ring trade and it is connected to the fee token.
+    pub fn is_overlapping(&self) -> bool {
+        let fee_token = match self
+            .projection
+            .node_indices()
+            .find(|i| self.projection[*i] == 0)
+        {
+            Some(token) => token,
+            None => {
+                // NOTE: The fee token is not in the graph, this means there are
+                // no orders buying or selling the fee token and therefore it
+                // cannot be overlapping.
+                return false;
+            }
+        };
+
+        petgraph::algo::bellman_ford(&self.projection, fee_token).is_err()
+    }
+}
+
+/// Type definition for a mapping of orders between buy and sell tokens.
+type OrderMap = HashMap<TokenId, HashMap<TokenId, Vec<Order>>>;
+
+/// A single order with a reference to the user.
+///
+/// Note that we approximate amounts and prices with floating point numbers.
+/// While this can lead to rounding errors it greatly simplifies the graph
+/// computations and still leads to acceptable estimates.
+#[cfg_attr(test, derive(Debug, PartialEq))]
+struct Order {
+    /// The user owning the order.
+    pub user: UserId,
+    /// The maximum capacity for this order, this is equivalent to the order's
+    /// sell amount (or price denominator). Note that orders are also limited by
+    /// their user's balance.
+    pub amount: f64,
+    /// The effective sell price for this order, that is price for this order
+    /// including fees of the sell token expressed in the buy token.
+    ///
+    /// Specifically, this is the sell amount over the buy amount or price
+    /// denominator over price numerator, which is the inverse price as
+    /// expressed by the exchange.
+    pub price: f64,
+    /// The weight of the order in the graph. This is the base-2 logarithm of
+    /// the price including fees. This enables transitive prices to be computed
+    /// using addition.
+    pub weight: f64,
```

Comment:
Should we just compute this on the fly when needed instead of storing a derivative field?

Reply:
Good question. I suppose it depends on how many times these values will be accessed relative to the complexity of evaluation. I tried looking for the complexity here https://en.wikipedia.org/wiki/Binary_logarithm#Computational_complexity 

but it doesn't appear to be explicitly stated.

Intuitively, I would say, if we are going to save on computation using addition here (as opposed to multiplication) then we might find the combined complexity of computing the log followed by addition at each access might run us into the same complexity of simply multiplying in the first place. With this in mind, I might be inclined to think that we only gain any advantage by computing the log once and storing it here to be used in future operations. Same argument would also imply that if we only access this stored value once then we haven't really gained anything.

I could get more precise about this intuition if necessary, but first I think we should determine approximately how often these values will be accessed.

Reply:
> I could get more precise about this intuition if necessary, but first I think we should determine approximately how often these values will be accessed.

I have benchmarks, I think the most reliable way to make optimizations is to benchmark against actual orderbooks. I'll make computed on the fly for now, and optimize later.